src/python/pants/core/goals/fmt.py

```diff
@@ -130,23 +143,80 @@ def files(self) -> tuple[str, ...]:
             return self.elements
 
     @classmethod
-    def _get_registration_rules(cls) -> Iterable[UnionRule]:
-        yield from super()._get_registration_rules()
+    def _get_registration_rules(cls, *, partitioner_type: PartitionerType) -> Iterable[UnionRule]:
+        yield from super()._get_registration_rules(partitioner_type=partitioner_type)
         yield UnionRule(FmtRequest, cls)
         yield UnionRule(FmtRequest.SubPartition, cls.SubPartition)
 
 
 class FmtTargetsRequest(FmtRequest, LintTargetsRequest):
+    @memoized_classproperty
+    def _default_single_partition_partitioner_rules(cls) -> Iterable:
+        # NB: This only works if the FieldSet has a single `SourcesField` field. We check here for
+        # a better user experience.
+        sources_field_name = None
+        for fieldname, fieldtype in _get_field_set_fields(cls.field_set_type).items():
+            if issubclass(fieldtype, SourcesField):
+                if sources_field_name is None:
+                    sources_field_name = fieldname
+                    break
+                raise TypeError(
+                    f"Type {cls.field_set_type} has multiple `SourcesField` fields."
+                    + " Pants can't provide a default partitioner."
+                )
+        else:
+            raise TypeError(
+                f"Type {cls.field_set_type} has does not have a `SourcesField` field."
+                + " Pants can't provide a default partitioner."
+            )
+
+        @rule(
+            _param_type_overrides={
+                "request": cls.PartitionRequest,
+                "subsystem": cls.tool_subsystem,
+            }
+        )
+        async def default_single_partition_partitioner(
+            request: FmtTargetsRequest.PartitionRequest, subsystem: SkippableSubsystem
```

Comment:
What happens if you do:
```python
        @rule
        async def default_single_partition_partitioner(
            request: cls.PartitionRequest, subsystem: cls.tool_subsystem
        ) -> Partitions: ...
```

without the param type overrides?

(Not questioning the need for this, just trying to understand why it is needed)

Reply:
The engine goes kaboom because the type annotation isn't runtime friendly per-se (it's complicated) and mypy absolutely goes apeshit because that's totally not a valid type lol.

Reply:
Well, oh, right now mypy does static analysis, which probably would be hard to reason about what that would become, yeah. About the type annotation not being runtime friendly, do you mean that it is a string value that you would have to translate using the scope where it was defined, and normally that string value is the key to the type in the scope, where as in this case, you would have to recurse one more time doing the lookup? (I think I get it, even though that last sentence feels anything but) :P

Reply:
Thank you for elaborating on that. I was aware of the future annotations feature toggle, but not really the unavailability of the `cls` symbol.

Reply:
No worries, 

it was a fun dive when I originally started looking into this pattern. If `cls` was closed it actually could've worked, and I very briefly thought about just making a pointless `cls` statement in the body to make it closed, but thought better of it :sweat_smile: